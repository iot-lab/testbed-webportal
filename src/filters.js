import Vue from 'vue'
import moment from 'moment'
import humanizeDuration from 'humanize-duration'

Vue.filter('fromTimestamp', function (value) {
  if (value) {
    return moment.unix(value).format('YYYY-MM-DD HH:mm:ss')
  }
})

Vue.filter('formatDate', function (value) {
  if (value) {
    return moment(String(value)).format('YYYY-MM-DD')
  }
})

Vue.filter('formatDateTime', function (value) {
  if (value) {
    return moment(String(value)).format('YYYY-MM-DD HH:mm')
  }
})

Vue.filter('humanizeDuration', function (value) {
  if (value) {
    let ms = value * 60 * 1000
    return humanizeDuration(ms, {
      units: ['y', 'd', 'h', 'm'],
      largest: 2,
      round: true,
    })
  }
})

Vue.filter('stripDomain', function (value) {
  if (value) {
    return value.replace('.iot-lab.info', '')
  }
})

Vue.filter('stateBadgeClass', function (value) {
  if (value) {
    switch (value) {
      case 'Alive':
      case 'Running':
        return 'badge-success'
      case 'Error':
      case 'toError':
      case 'Absent':
        return 'badge-danger'
      case 'Waiting':
      case 'Hold':
      case 'Finishing':
      case 'Busy':
        return 'badge-warning'
      case 'Resuming':
      case 'Launching':
      case 'toLaunch':
      case 'toAckReservation':
        return 'badge-info'
      case 'Suspected':
      case 'Suspended':
      case 'Stopped':
        return 'badge-dark'
      case 'Terminated':
      default:
        return 'badge-secondary'
    }
  }
})

Vue.filter('formatArchi', function (value) {
  if (value) {
    return value.split(':')[0].replace('_', ' ')
  }
})

Vue.filter('formatArchiRadio', function (value) {
  if (value) {
    if (value.split(':')[1]) {
      return value.split(':')[0].replace('_', ' ') + ' (' + value.split(':')[1] + ')'
    }
    return value.split(':')[0].replace('_', ' ')
  }
})
