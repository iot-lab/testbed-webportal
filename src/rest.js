import axios from 'axios'

const API_URL = `https://${process.env.VUE_APP_IOTLAB_HOST}/api/`

export const iotlab = {

  api: axios.create({baseURL: API_URL, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')}),

  create (username, password) {
    this.api = axios.create({
      baseURL: API_URL,
      auth: {
        username: username,
        password: password,
      },
    })
  },

  // USERS API

  async signup (user, nbUsers) {
    if (nbUsers) {
      // create multiple user accounts
      await iotlab.api.post(`/users?nbusers=${nbUsers}`, user)
    } else {
      // create single user account
      await iotlab.api.post('/users', user)
    }
  },

  async activateAccount (hash) {
    await iotlab.api.post('/users/activate/', {'hash': hash})
  },

  async deleteUser (login = undefined) {
    if (login) {
      await iotlab.api.delete(`/users/${login}`)
    } else {
      await iotlab.api.delete(`/user`)
    }
  },

  async resetPassword (email) {
    await iotlab.api.post(`/users/reset_password`, {'email': email})
  },

  async changePassword (oldPwd, newPwd, confirmPwd) {
    await iotlab.api.put('/user/password', {
      'old_password': oldPwd,
      'new_password': newPwd,
      'confirm_new_password': confirmPwd,
    })
  },

  async getSSHkeys () {
    return iotlab.api.get('/user/keys').then(resp => resp.data.sshkeys)
  },

  async modifySSHkeys (keys) {
    return iotlab.api.post('/user/keys', {'sshkeys': keys})
  },

  async deleteSSHkey (id) {
    return iotlab.api.delete(`/user/keys/${id}`)
  },

  async getUsers ({status, group, search} = {}) {
    let params = {}
    if (status) {
      params.status = status
    }
    if (group) {
      params.group = group
    }
    if (search) {
      params.search = search
    }
    return iotlab.api.get('/users', {params: params}).then(resp => resp.data)
  },

  async getUserInfo (login = undefined) {
    if (login) {
      return iotlab.api.get(`/users/${login}`).then(resp => resp.data)
    } else {
      return iotlab.api.get('/user').then(resp => resp.data)
    }
  },

  async setUserInfo (user, login = undefined) {
    if (login) {
      return iotlab.api.put(`/users/${login}`, user)
    } else {
      return iotlab.api.put('/user', user)
    }
  },

  async activateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'active'})
  },

  async deactivateUser (login) {
    return iotlab.api.put(`/users/${login}`, {'status': 'inactive'})
  },

  async setUserGroups (login, groups) {
    return iotlab.api.put(`/users/${login}`, {'groups': groups})
  },

  async sendMail (recipients, subject, message) {
    await iotlab.api.post('/users/email', {
      recipients: recipients,
      subject: subject,
      body: message,
    })
  },

  // EXPERIMENT API

  async getUserExperimentsCount () {
    return iotlab.api.get('/experiments/total').then(resp => resp.data)
  },

  async getAllExperimentsCount (user) {
    if (user) {
      return iotlab.api.get(`/experiments/total/users/${user}`).then(resp => resp.data)
    } else {
      return iotlab.api.get(`/experiments/total/all`).then(resp => resp.data)
    }
  },

  async getAllExperiments ({limit, offset, user, state = 'Terminated,Error,Running,Finishing,Resuming,toError,Waiting,Launching,Hold,toLaunch,toAckReservation,Suspended'} = {}) {
    let params = {}
    if (state) {
      params.state = state
    }
    if (limit) {
      params.limit = Math.min(limit, 500) // IoT-LAB api restrict queries to 500 entries
    }
    if (offset) {
      params.offset = offset
    }
    if (user === '@self') {
      return iotlab.api.get('/experiments', {params: params}).then(resp => resp.data.items)
    } else if (user) {
      return iotlab.api.get(`/experiments/users/${user}`, {params: params}).then(resp => resp.data.items)
    } else {
      return iotlab.api.get('/experiments/all', {params: params}).then(resp => resp.data.items)
    }
  },

  async getRunningExperiments () {
    return iotlab.api.get('/experiments/running').then(resp => resp.data.items)
  },

  async getExperiment (id) {
    return iotlab.api.get(`/experiments/${id}`).then(resp => resp.data)
  },

  async getExperimentArchive (id) {
    return iotlab.api.get(`/experiments/${id}/data`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async getExperimentDeployment (id) {
    return iotlab.api.get(`/experiments/${id}/deployment`).then(resp => resp.data)
  },

  async stopExperiment (id) {
    return iotlab.api.delete(`/experiments/${id}`)
  },

  async submitExperiment ({type, name, duration, reservation, nodes, firmwareassociations, firmwares, profileassociations, profiles} = {}) {
    const formData = new FormData()
    formData.append('experiment.json', JSON.stringify({
      type: type,
      duration: duration,
      reservation: reservation,
      name: name,
      nodes: nodes,
      firmwareassociations: firmwareassociations,
      profileassociations: profileassociations,
    }))
    // attached firmware names should be unique
    let attachedFirmwares = []
    for (let asso of firmwareassociations || []) {
      if (asso.firmwarename in firmwares && !attachedFirmwares.includes(asso.firmwarename)) {
        formData.append(asso.firmwarename, new Blob([firmwares[asso.firmwarename]], {type: 'application/octet-stream'}), asso.firmwarename)
        attachedFirmwares.push(asso.firmwarename)
      }
    }
    return iotlab.api.post('/experiments', formData).then(resp => resp.data)
  },

  async checkFirmware (fileString) {
    const formData = new FormData()
    formData.append('firmware.bin', fileString)
    return iotlab.api.post('/firmwares/checker', formData).then(resp => resp.data)
  },

  async flashFirmware (id, nodes, fileString) {
    const formData = new FormData()
    formData.append(`experiment${id}.json`, JSON.stringify(nodes))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/experiments/${id}/nodes/flash`, formData).then(resp => resp.data)
  },

  async flashResourcesFirmware (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/flash/${name}`, nodes).then(resp => resp.data)
  },

  async updateMonitoring (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/nodes/monitoring/${name}`, nodes).then(resp => resp.data)
  },

  async updateExperimentMobilityCircuit (id, nodes, name) {
    return iotlab.api.post(`/experiments/${id}/robots/mobility/${name}`, nodes).then(resp => resp.data)
  },

  async reloadExperiment (id) {
    return iotlab.api.post(`/experiments/${id}/reload`, {}).then(resp => resp.data)
  },

  // USER GROUPS

  async getUserGroups () {
    return iotlab.api.get('/groups').then(resp => resp.data)
  },

  async createGroup (name) {
    return iotlab.api.post(`/groups`, {name: name})
  },

  async deleteGroup (group) {
    return iotlab.api.delete(`/groups/${group}`)
  },

  // OTHER API

  async getSites () {
    return iotlab.api.get('/sites').then(resp => resp.data.items)
  },

  async getSitesDetails () {
    return iotlab.api.get('/sites/details').then(resp => resp.data.items)
  },

  async getNodes () {
    return iotlab.api.get('/nodes').then(resp => resp.data.items)
  },

  async updateNodesProperties () {
    return iotlab.api.put('/nodes')
  },

  async updateNodesPropertiesOldApi () {
    const OLD_API = `https://${process.env.VUE_APP_IOTLAB_HOST}/rest/`
    const oldApi = axios.create({baseURL: OLD_API, auth: JSON.parse(localStorage.getItem('apiAuth') || '{}')})
    return oldApi.post('/admin/nodes')
  },

  async getNodesIds () {
    return iotlab.api.get('/nodes/ids').then(resp => resp.data.items)
  },

  async getExperimentNodes (id) {
    return iotlab.api.get(`/experiments/${id}/nodes`).then(resp => resp.data.items)
  },

  async getExperimentToken (id) {
    return iotlab.api.get(`/experiments/${id}/token`).then(resp => resp.data.token)
  },

  async sendNodesCommand (id, cmd, nodes) {
    // cmd in [start, stop, reset, flash-idle, profile-reset, debug-start, debug-stop]
    return iotlab.api.post(`/experiments/${id}/nodes/${cmd}`, nodes).then(resp => resp.data)
  },

  // STORES (firmwares, profiles, etc)

  async getMobilityCircuits () {
    return iotlab.api.get('/mobilities/circuits').then(resp => resp.data.items)
  },

  async getMobilityCircuit (name) {
    return iotlab.api.get(`/mobilities/circuits/${name}`).then(resp => resp.data)
  },

  async getRobotsSiteMapImage (site) {
    return iotlab.api.get(`/robots/${site}/map/image`, {
      responseType: 'arraybuffer',
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    }).then(resp => Buffer.from(resp.data, 'binary').toString('base64')
    )
  },

  async getRobotsSiteMapConfig (site) {
    return iotlab.api.get(`/robots/${site}/map/config`).then(resp => resp.data)
  },

  async createMobilityCircuit (circuit) {
    return iotlab.api.post(`/mobilities/circuits`, circuit)
  },

  async updateMobilityCircuit (name, circuit) {
    return iotlab.api.put(`/mobilities/circuits/${name}`, circuit)
  },

  async deleteMobilityCircuit (name) {
    return iotlab.api.delete(`/mobilities/circuits/${name}`)
  },

  async getMonitoringProfiles () {
    return iotlab.api.get('/monitoring').then(resp => resp.data)
  },

  async getMonitoringProfile (name) {
    return iotlab.api.get(`/monitoring/${name}`).then(resp => resp.data)
  },

  async updateMonitoringProfile (name, profile) {
    return iotlab.api.put(`/monitoring/${name}`, profile)
  },

  async createMonitoringProfile (profile) {
    return iotlab.api.post(`/monitoring`, profile)
  },

  async deleteMonitoringProfile (name) {
    return iotlab.api.delete(`/monitoring/${name}`)
  },

  async getFirmwares () {
    return iotlab.api.get('/firmwares').then(resp => resp.data)
  },

  async getFirmware (name) {
    return iotlab.api.get(`/firmwares/${name}`).then(resp => resp.data)
  },

  async getFirmwareFile (name) {
    return iotlab.api.get(`/firmwares/${name}/file`, {responseType: 'arraybuffer'}).then(resp => resp.data)
  },

  async updateFirmware (name, firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    if (fileString) {
      formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    }
    return iotlab.api.put(`/firmwares/${name}`, formData).then(resp => resp.data)
  },

  async createFirmware (firmware, fileString) {
    const formData = new FormData()
    formData.append(`firmware.json`, JSON.stringify(firmware))
    formData.append('firmware.bin', new Blob([fileString], {type: 'application/octet-stream'}), 'firmware.bin')
    return iotlab.api.post(`/firmwares`, formData).then(resp => resp.data)
  },

  async deleteFirmware (name) {
    return iotlab.api.delete(`/firmwares/${name}`)
  },
}
