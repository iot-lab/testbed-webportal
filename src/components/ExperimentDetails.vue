<template>
  <div class="container mt-3">
    <h2>
      Experiment <span class="text-muted">{{experiment.name}} <small>#{{id}}</small></span>
    </h2>
    <ul class="list-unstyled">
      <li>User <b>{{experiment.user}}</b></li>
      <li>Name <b>{{experiment.name}}</b></li>
      <li>Duration <b>{{experiment.duration | humanizeDuration}}</b></li>
      <li>Number of nodes <b>{{experiment.nodes.length}}</b></li>
      <li>Type <b>{{experiment.type}}</b></li>
      <li>State <span class="badge" :class="experiment.state | stateBadgeClass">{{experiment.state}}</span></li>
    </ul>
    <a href="" class="btn btn-sm btn-outline-danger mb-3" @click.prevent="delExperiment(id)"
      v-if="states.scheduled.includes(experiment.state)">
      <i class="fa fa-ban"></i> Stop
    </a>
    <a href="" class="btn btn-sm btn-outline-secondary mb-3" @click.prevent="reloadExperiment(id)">
      <i class="fa fa-refresh"></i> Restart
    </a>
    <a href="" class="btn btn-sm btn-outline-secondary mb-3" @click.prevent="this.alert('todo')">
      <i class="fa fa-clone"></i> Clone
    </a>

    <table class="table table-striped table-sm">
      <thead>
        <tr>
          <th>Nodes</th>
          <th>Firmware</th>
          <th>Monitoring</th>
          <th>Mobility</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="node in experiment.nodes">
          <td>{{node}}</td>
          <td>{{getFirmware(node)}}</td>
          <td>{{getMonitoring(node)}}</td>
          <td>{{getMobility(node)}}</td>
        </tr>
      </tbody>
    </table>
  </div> <!-- container -->
</template>

<script>
import { iotlab } from '@/rest'
import { experimentStates } from '@/assets/js/iotlab-utils'

export default {
  name: 'ExperimentDetails',

  props: {
    id: {
      type: Number,
    },
  },

  data () {
    return {
      experiment: undefined,
      states: experimentStates,
    }
  },

  async created () {
    try {
      this.experiment = await iotlab.getExperiment(this.id)
    } catch (err) {
      this.$notify({text: err.message, type: 'error'})
    }
  },

  methods: {
    getFirmware (node) {
      if (!this.experiment.firmwareassociations) return
      return this.experiment.firmwareassociations.reduce((acc, asso) => (asso.nodes.includes(node) ? asso.firmwarename : acc), '')
    },
    getMobility (node) {
      if (!this.experiment.associations || !this.experiment.associations.mobility) return
      return this.experiment.associations.mobility.reduce((acc, asso) => (asso.nodes.includes(node) ? asso.mobilityname : acc), '')
    },
    getMonitoring (node) {
      return '?'
    },
    async delExperiment (id) {
      try {
        await iotlab.stopExperiment(id)
        this.$notify({text: `Experiment ${id} stopped`, type: 'success'})
        this.experiment.state = 'Finishing'
      } catch (err) {
        this.$notify({text: err.message, type: 'error'})
      }
    },
    async reloadExperiment (id) {
      try {
        let newExp = await iotlab.reloadExperiment(id)
        this.$notify({text: `Experiment ${newExp.id} submitted`, type: 'success'})
      } catch (err) {
        this.$notify({text: err.message, type: 'error'})
      }
    },

  },
}
</script>

<style>

</style>
