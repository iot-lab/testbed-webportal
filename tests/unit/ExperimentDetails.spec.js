import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Notifications from 'vue-notification'
import ExperimentDetails from '@/views/ExperimentDetails'
import { iotlab } from '@/rest'
import * as Filters from '@/filters'

jest.mock('@/rest')
iotlab.getExperiment = jest.fn().mockResolvedValue({"submitted_duration":20,"nb_nodes":2,"stop_date":"2017-01-16T14:22:41Z","type":"physical","firmwareassociations":[{"nodes":["m3-12.saclay.iot-lab.info","m3-9.saclay.iot-lab.info"],"firmwarename":"tutorial_m3.elf"}],"scheduled_date":"1970-01-01T00:00:00Z","submission_date":"2017-01-16T14:02:21Z","nodes":["m3-9.saclay.iot-lab.info","m3-12.saclay.iot-lab.info"],"effective_duration":20,"name":"test1","id":58099,"state":"Terminated","user":"formet","start_date":"2017-01-16T14:02:23Z"})
iotlab.getExperimentDeployment = jest.fn().mockResolvedValue({"0":["m3-9.saclay.iot-lab.info","m3-12.saclay.iot-lab.info"]})

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(Notifications)
const router = new VueRouter()

localVue.filter('formatDateTimeSec', Filters.formatDateTimeSec)
localVue.filter('humanizeDuration', Filters.humanDuration)
localVue.filter('stateBadgeClass', Filters.stateBadgeClass)

describe('ExperimentDetails.vue', () => {

  it('should match snapshot', async () => {
    const wrapper = await shallowMount(ExperimentDetails, {localVue, router, propsData: {
      id: 58099,
    }})
    await wrapper.vm.$nextTick() // wait for promise to resolve
    await wrapper.vm.$nextTick() // wait for promise to resolve
    await wrapper.vm.$nextTick() // wait for promise to resolve
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

})
