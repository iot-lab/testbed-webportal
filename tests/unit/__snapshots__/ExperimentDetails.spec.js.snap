// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentDetails.vue should match snapshot 1`] = `
<div
  class="container mt-3"
>
  <h2>
    
    Experiment 
    <span
      class="text-muted"
    >
      test1 
      <small>
        #58099
      </small>
    </span>
  </h2>
   
  <ul
    class="list-unstyled"
  >
    <li>
      User 
      <b>
        formet
      </b>
    </li>
     
    <li>
      Submitted 
      <b>
        2017-01-16 15:02:21
      </b>
    </li>
     
    <!---->
     
    <li>
      Started 
      <b>
        2017-01-16 15:02:23
      </b>
    </li>
     
    <li>
      Stopped 
      <b>
        2017-01-16 15:22:41
      </b>
    </li>
     
    <li>
      
      Duration 
      <b>
        20 minutes
      </b>
       
      <!---->
    </li>
     
    <!---->
     
    <li>
      Nodes 
      <b>
        2
      </b>
    </li>
     
    <li>
      State 
      <span
        class="badge badge-secondary"
      >
        Terminated
      </span>
    </li>
  </ul>
   
  <!---->
   
  <table
    class="table table-striped table-sm"
  >
    <thead>
      <tr>
        <!---->
         
        <th>
          Nodes
        </th>
         
        <th>
          UID
        </th>
         
        <th>
          Firmware
        </th>
         
        <th>
          Monitoring
        </th>
         
        <th
          class="text-center"
        >
          Deployment
        </th>
         
        <!---->
      </tr>
    </thead>
     
    <tbody>
      <tr
        class=""
      >
        <!---->
         
        <td>
          m3-9.saclay.iot-lab.info
        </td>
         
        <td>
          
        </td>
         
        <td>
          tutorial_m3.elf
        </td>
         
        <td>
          
        </td>
         
        <td
          class="text-center"
        >
          
        </td>
         
        <!---->
      </tr>
       
      <tr>
        <td
          class="p-0"
          colspan="7"
        >
          <div
            class="d-flex align-items-center justify-content-between bg-secondary"
          >
            <terminal-stub
              cols="80"
              expid="58099"
              node="m3-9.saclay.iot-lab.info"
              rows="20"
              user=""
            />
             
            <!---->
          </div>
        </td>
      </tr>
      <tr
        class=""
      >
        <!---->
         
        <td>
          m3-12.saclay.iot-lab.info
        </td>
         
        <td>
          
        </td>
         
        <td>
          tutorial_m3.elf
        </td>
         
        <td>
          
        </td>
         
        <td
          class="text-center"
        >
          
        </td>
         
        <!---->
      </tr>
       
      <tr>
        <td
          class="p-0"
          colspan="7"
        >
          <div
            class="d-flex align-items-center justify-content-between bg-secondary"
          >
            <terminal-stub
              cols="80"
              expid="58099"
              node="m3-12.saclay.iot-lab.info"
              rows="20"
              user=""
            />
             
            <!---->
          </div>
        </td>
      </tr>
    </tbody>
  </table>
   
  <div
    aria-hidden="true"
    aria-labelledby="firmwareModal"
    class="modal fade firmware-modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header bg-light"
        >
          <h5
            class="modal-title"
          >
            Select firmware
          </h5>
           
          <button
            aria-label="Close"
            class="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
         
        <div
          class="modal-body py-4"
        >
          <label
            class="custom-file"
          >
            <input
              class="custom-file-input"
              id="file"
              type="file"
            />
             
            <span
              class="custom-file-control"
            >
              
            </span>
          </label>
           
          <hr />
           
          <firmwarelist-stub
            archi="m3,"
            select="true"
          />
        </div>
         
        <div
          class="modal-footer dborder-0 dbg-light"
        >
          <button
            class="btn"
            data-dismiss="modal"
            type="button"
          >
            Close
          </button>
           
          <button
            class="btn btn-primary"
            type="button"
          >
            Flash
          </button>
        </div>
      </div>
    </div>
  </div>
   
  <div
    aria-hidden="true"
    aria-labelledby="monitoringModal"
    class="modal fade monitoring-modal"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header bg-light"
        >
          <h5
            class="modal-title"
          >
            Select monitoring
          </h5>
           
          <button
            aria-label="Close"
            class="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </div>
         
        <div
          class="modal-body py-4"
        >
          <monitoringlist-stub
            archi="m3"
            select="true"
          />
        </div>
         
        <div
          class="modal-footer dborder-0 dbg-light"
        >
          <button
            class="btn"
            data-dismiss="modal"
            type="button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
